[tutorial] new tutorial extensions (#4876)
withastro/docs: src/content/docs/en/extend/add-view-transitions.mdx

    ```astro title="src/layouts/BaseLayout.astro" ins={2,15}
    ---
    import { ViewTransitions } from 'astro:transitions';
    import Header from "../components/Header.astro";
    import Footer from "../components/Footer.astro";
    import "../styles/global.css";
    const { pageTitle } = Astro.props;
    ---
    <html lang="en">
      <head>
        <meta charset="utf-8" />
        <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
        <meta name="viewport" content="width=device-width" />
        <meta name="generator" content={Astro.generator} />
        <title>{pageTitle}</title>
        <ViewTransitions />
      </head>
      <body>
        <Header />
        <h1>{pageTitle}</h1>
        <slot />
        <Footer />
        <script>
          import "../scripts/menu.js";
        </script>
      </body>
    </html>
    ```

    No other configuration is necessary to enable Astro's default client-side navigation! Astro will create default page animations based on the similarities between the old and new page, and will also provide fallback behavior for unsupported browsers.

> 3. Navigate between pages in your site preview
> 
>     As you move between pages on your site, notice that the old page content appears to fade out as the new page content fades in. You can now add custom behavior to control individual page links if you are not satisfied with the defaults.

Reviewer: Add note about mobile view, or generally that scripts don't work
```suggestion
3. Navigate between pages in your site preview

    As you move between pages on your site, notice that the old page content appears to fade out as the new page content fades in. You can now add custom behavior to control individual page links if you are not satisfied with the defaults.
```

### Update scripts

With view transitions, some scripts may no longer re-run after page navigation like they do with full-page browser refreshes. There are two events during client-side routing that you can listen for, and fire events when they occur: [`astro:page-load`](/en/guides/view-transitions/#astropage-load) and [`astro:after-swap`](/en/guides/view-transitions/#astroafter-swap).

4. Make the script controling the `<Hamburger />` mobile menu component available after navigating to a new page.

     To make your mobile menu interactive after navigating to a new page, add the following code that listens for the `astro:page-load` event which runs at the end of page navigation, and in response, runs the existing script to make the hamburger menu function when clicked:

    ```js title="src/scripts/menu.js" ins={1,5}
    document.addEventListener('astro:page-load', () => {
      document.querySelector('.hamburger').addEventListener('click', () => {
        document.querySelector('.nav-links').classList.toggle('expanded');
      });
    });
    ```

5. Make the script controlling the theme toggle available after page navigation.

>     The `<script>` that controls the light/dark theme toggle is located in the `<ThemeIcon />` component. For the theme toggle to continue to function on every page, remove the `is:inline` attribute from the script add the same event listener as in the previous example so that `astro:page-load` event can trigger your existing function.

Reviewer 2: ```suggestion
    The `<script>` that controls the light/dark theme toggle is located in the `<ThemeIcon />` component. For the theme toggle to continue to function on every page, remove the `is:inline` attribute from the script and add the same event listener as in the previous example so that `astro:page-load` event can trigger your existing function.
```

I think this may be missing an "and" or a comma between the 2 actions.

    
    Update the existing script tag so that your function runs in response to the `astro:page-load` event, making your theme toggle interactive after the new page is fully-loaded and visible to the user:

    ```astro title="src/components/ThemeIcon.astro" ins={8,38} del="is:inline"
    ---
    ---

>     <button id="themeToggle"> ... </button>
> 
>     <style> ... </style>

Reviewer 2: ```suggestion
    <button id="themeToggle"> /* ... */ </button>

    <style> /* ... */ </style>
```

This avoid breaking syntax highlighting of the JavaScript below.

    ```astro title="src/layouts/BaseLayout.astro" ins='transition:animate="slide"'
    ---
    import Header from "../components/Header.astro";
    import Footer from "../components/Footer.astro";
    import "../styles/global.css";
    import { ViewTransitions } from 'astro:transitions';
    const { pageTitle } = Astro.props;
    ---

    <html lang="en">
      <head>
        <meta charset="utf-8" />
        <link rel="icon" type="image/svg+xml" href="/favicon.svg" />
        <meta name="viewport" content="width=device-width" />
        <meta name="generator" content={Astro.generator} />
        <title>{pageTitle}</title>
        <ViewTransitions />
      </head>
      <body>
        <Header />
        <h1 transition:animate="slide">{pageTitle}</h1>
        <slot />
        <Footer />
        <script>
          import "../scripts/menu.js";
        </script>
      </body>
    </html>
    ```

    In your browser preview, you will now see the page titles slide on to the screen, while other elements such as the body text continues to fade in and out. 

    <Box icon="puzzle-piece">
      ### Try it yourself - Make the navigation links slide in

>       Add an animation directive to make the `<div>` containing all the header links slide in on page navigation, following the [same steps as above](#customize-transition-animations).

Reviewer: Specified `Navigation.astro` to guide people towards the component. Maybe the sentence can be reworked a little too :thinking: 
```suggestion
      Add an animation directive to make the `<div>` in `Navigation.astro` containing all the header links slide in on page navigation, following the [same steps as above](#customize-transition-animations).
```

    Sometimes you will want a full browser reload when visitors click a certain link. For example, you may be linking to a page that does not also use the `<ViewTransitions />` router, or to a file directly such as a `.pdf`. 

    To make it so that your browser refreshes every time you click the navigation link to go to your About page, add the `data-astro-reload` attribute to the `<a>` tag in your `<Navigation/>` component. This will override the `<ViewTransitions />` router entirely, and any of the view transition animations, for this one link.

     ```astro title="src/components/Navigation.astro" ins='data-astro-reload'
      ---
      ---
      <div transition:animate="slide" class="nav-links">
        <a href="/">Home</a>
        <a href="/about/" data-astro-reload >About</a>
        <a href="/blog/">Blog</a>
        <a href="/tags/" >Tags</a>
      </div>
      ```
    
    Now, when you click the navigation link to your About page, no animations will occur. The page links and title will not slide in, and the page content will not fade in when you navigate to your About page **using this link**.

>  10. Add a link to your About page from your author name in your Markdown layout for blog posts   

Reviewer 3: ```suggestion
10. Add a link to your About page from your author name in your Markdown layout for blog posts   
```

Might actually solve the indentation issue, actually... maybe?

---
title: Tutorial - Extend with View Transitions
description: >-
  Add View Transitions to the Build a Blog tutorial code.
i18nReady: true
---
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import FileTree from '~/components/FileTree.astro';
import Since from '~/components/Since.astro';
import TypeScriptSettingTabs from '~/components/tabs/TypeScriptSettingTabs.astro';
import RecipeLinks from '~/components/RecipeLinks.astro';
import Box from '~/components/tutorial/Box.astro';
import MultipleChoice from '~/components/tutorial/MultipleChoice.astro';
import Spoiler from '~/components/Spoiler.astro';

> **View transitions** are a way to control what happens when visitors navigate between pages on your site. Astro's View Transitions API allows you to add optional navigation features including smooth page transitions and animations, controling the brower's history stack of visited pages and preventing full-page refreshes in order to persist some page elements and state while updating the content displayed.

Reviewer 4: two typos and one comma
```suggestion
**View transitions** are a way to control what happens when visitors navigate between pages on your site. Astro's View Transitions API allows you to add optional navigation features including smooth page transitions and animations, controlling the browser's history stack of visited pages, and preventing full-page refreshes in order to persist some page elements and state while updating the content displayed.
```

4. Make the script controling the `<Hamburger />` mobile menu component available after navigating to a new page.

     To make your mobile menu interactive after navigating to a new page, add the following code that listens for the `astro:page-load` event which runs at the end of page navigation, and in response, runs the existing script to make the hamburger menu function when clicked:

    ```js title="src/scripts/menu.js" ins={1,5}
    document.addEventListener('astro:page-load', () => {
      document.querySelector('.hamburger').addEventListener('click', () => {
        document.querySelector('.nav-links').classList.toggle('expanded');
      });
    });
    ```

5. Make the script controlling the theme toggle available after page navigation.

    The `<script>` that controls the light/dark theme toggle is located in the `<ThemeIcon />` component. For the theme toggle to continue to function on every page, remove the `is:inline` attribute from the script and add the same event listener as in the previous example so that `astro:page-load` event can trigger your existing function.
    

>     Update the existing script tag so that your function runs in response to the `astro:page-load` event, making your theme toggle interactive after the new page is fully-loaded and visible to the user:

Reviewer 4: I believe "fully loaded" here is without a hyphen.
```suggestion
    Update the existing script tag so that your function runs in response to the `astro:page-load` event, making your theme toggle interactive after the new page is fully loaded and visible to the user:
```

    Add this new script to the `<ThemeIcon />` component, in addition to the script that controls the theme toggle.

    ```astro title="src/components/ThemeIcon.astro" ins={3-9}
    <script> ... <script>

    <script>
      document.addEventListener('astro:after-swap', () => {
        localStorage.theme === 'dark' 
        ? document.documentElement.classList.add("dark")
        : document.documentElement.classList.add("light");
      }); 
    </script>
    ```

    Now, every page change **that uses the `<ViewTransitions />` router for client-side navigation** (and therefore access to the `astro:after-swap` event) will be able to detect `theme: dark` from the browser's `localStorage` and update the current page accordingly before the page is rendered for the viewer.

### Customize Transition animations

7. Change the default `fade` animation to a `slide` for the page title.

    With view transitions enabled, you currently have a small fade in and out set for all page transition animations. Astro also provides a built-in `slide` animation. To change the type of animation for a single element, add the `transition:animation=""` directive.

>     For example, to make the page titles slide in instead of fade in, add `transition:animate="slide"` to the `<h1>` element in the BaseLayout:

Reviewer 4: There are several places here we are using "fade in", maybe with the exception of the first use (paragraph above), I believe all these should be "fade-in" (with a hyphen) instead.

Author: I think I'm using it as a verb everywhere, not a noun, so I feel pretty confident these are correct as is. (Admittedly there may be one or two places where either verb or noun could be correct, but I checked them all and there isn't anywhere I'd clearly think "noun")

 
There are times when you will want or need a full-page browser refresh. For example, when a link takes a visitor to a `.pdf` document, you will need the browser to load that new page from the server. Even with view transitions enabled in your Astro project, you will be able to specify how the browser should navigate both by default and on a per-link basis, even opting out of client-side routing entirely.

Read more about [Astro's view transitions](/en/guides/view-transitions/) in our guide, or get started with the instructions below to convert a basic blog from `src/pages/posts/` to `src/content/posts/`.

<Box icon="question-mark">
### Test your knowledge

1. Adding view transitions to my astro site...

    <MultipleChoice>
      <Option>
        Requires more than 2 lines of code to implement site-wide by default
      </Option>
      <Option isCorrect>

>         Changes the default type of page navigation on page that contains the `<ViewTransitions />` router in its `<head>`

Reviewer: ```suggestion
        Changes the default type of page navigation on a page that contains the `<ViewTransitions />` router in its `<head>`
```

    📚 Read more about the different [transition directives](/en/guides/view-transitions/#transition-directives) and [customizing animations](/en/guides/view-transitions/#customizing-animations).

### Force a full browser reload for some links

9. Prevent client-side routing and instead require the browser to reload when navigating to your About page.

    Sometimes you will want a full browser reload when visitors click a certain link. For example, you may be linking to a page that does not also use the `<ViewTransitions />` router, or to a file directly such as a `.pdf`. 

    To make it so that your browser refreshes every time you click the navigation link to go to your About page, add the `data-astro-reload` attribute to the `<a>` tag in your `<Navigation/>` component. This will override the `<ViewTransitions />` router entirely, and any of the view transition animations, for this one link.

     ```astro title="src/components/Navigation.astro" ins='data-astro-reload'
      ---
      ---
      <div transition:animate="slide" class="nav-links">
        <a href="/">Home</a>
        <a href="/about/" data-astro-reload >About</a>
        <a href="/blog/">Blog</a>

>         <a href="/tags/" >Tags</a>

Reviewer 5: ```suggestion
        <a href="/tags/">Tags</a>
```

Reviewer 5: I am sorry that I am nitting spaces, but i figured with this feature it isn't too bad of a thing...?

10. Add a link to your About page from your author name in your Markdown layout for blog posts   

    `data-astro-reload` only triggers a full browser refresh when going to a new page **from the link it is added to**. It does not control all instances of navigating to your About page.

    In your `<MarkdownPostLayout />` component, add a link to your About page on your author name:

    ```astro title="src/layouts/MarkdownPostLayout.astro" ins='<a href="/about/">' ins="</a>"
    ---
    import BaseLayout from "./BaseLayout.astro";
    import { fade } from 'astro:transitions';
    const { frontmatter } = Astro.props;
    ---

    <BaseLayout pageTitle={frontmatter.title}>
      <p>{frontmatter.pubDate.slice(0, 10)}</p>
      <p transition:animate={fade({ duration: '2s' })} ><em>{frontmatter.description}</em></p>
      <p>Written by: <a href="/about/">{frontmatter.author}</a></p>
      <img src={frontmatter.image.url} width="300" alt={frontmatter.image.alt} />
      <slot />
    </BaseLayout>
    ```

    If you visit any blog post in your browser preview, and then click on the linked author name to be taken to the About page, what does the page navigation look like?

    <p>

>       When a visitor clicks a link to the About page from an individual blog post, the page title and header navigation links<Spoiler>slide in across the screen</Spoiler> because <Spoiler>the `data-astro-reload` attribute is not set on these links</Spoiler>

Reviewer 5: ```suggestion
      When a visitor clicks a link to the About page from an individual blog post, the page title and header navigation links <Spoiler>slide in across the screen</Spoiler> because <Spoiler>the `data-astro-reload` attribute is not set on these links</Spoiler>
```
